hammers/scripts/orphans_detector.py

```diff
@@ -0,0 +1,173 @@
+        else:
+            message = []
+            if not user_enabled:
+                message.append('User {} is deactivated'.format(user_id))
+            if not project_enabled:
+                message.append('Project {} is deactivated'.format(project_id))
+            orphans[instance_id] = ' and '.join(message)
+    
+    return orphans
+    
+
+def generate_report(orphan_dict, title):
+    if not orphan_dict:
+        return None
+    
+    report = []
+    report.append(title)
+    report.append("{:<45} {:<50}".format('ID','Message'))
+    for k, v in orphan_dict.iteritems():
+        report.append("{:<45} {:<50}".format(k, v))
+    
+    return '\n'.join(report)
+    
+def main(argv=None):
+    if argv is None:
+        argv = sys.argv
+
+    parser = argparse.ArgumentParser(description='Detects orphan leases and remove them.')
+    
+    mysqlargs = MySqlArgs({
+        'user': 'root',
+        'password': '',
+        'host': 'localhost',
+        'port': 3306,
+    })
+    mysqlargs.inject(parser)
+
+    parser.add_argument('-d', '--dbversion', type=str,
+        help='Version of the database. Schemas differ, pick the appropriate one.',
+        choices=[query.LIBERTY, query.OCATA], default=query.OCATA)
+    parser.add_argument('--slack', type=str,
+        help='JSON file with Slack webhook information to send a notification to')
+    parser.add_argument('--kvm', help='Run at KVM site', action='store_true')
+
+    args = parser.parse_args(argv[1:])
+    mysqlargs.extract(args)
+    
+    db = mysqlargs.connect()
+    db.version = args.dbversion
+    
+    kvm = args.kvm
+    
+    if args.slack:
+        slack = Slackbot(args.slack, script_name='orphan-detector')
+    else:
+        slack = None
+
+    if kvm:
+        # at kvm site
+        auth = v2.Password(username=os.environ['OS_USERNAME'], 
+                           password=os.environ['OS_PASSWORD'], 
+                           tenant_name=os.environ['OS_TENANT_NAME'], 
+                           auth_url=os.environ['OS_AUTH_URL'])
+        sess = session.Session(auth=auth)
+        keystone = client.Client(session=sess)
+    
+        orphan_instances = get_orphan_instances_kvm(db, keystone)
+    else:
+        orphan_leases_report = generate_report(get_orphan_leases(db), "-" * 45 + "ORPHAN LEASES" + "-" * 45)
+        if slack:
+            if orphan_leases_report:
+                slack.post('orphan-detector', orphan_leases_report, color='#FF0000')
+            else:
+                slack.post('orphan-detector', 'No orphan leases detected', color='#000000')
+        else:
+            if orphan_leases_report:
+                print(orphan_leases_report)
+            else:
+                print('No orphan leases detected')
+                
+        orphan_instances = get_orphan_instances(db)
+    
+    orphan_instances_report = generate_report(orphan_instances, "-" * 45 + "ORPHAN INSTANCES" + "-" * 45)
```

Comment:
Is this doing the same job as lines 146-158? I think maybe you could change line 146 to simply be:

```
orphan_instances = get_orphan_instances(db)
```

And then you can generate the report and do the branching about where to send the report, as you have it.

Reply:
@diurnalist Not exactly the same job. One report for leases, and one for instances.

Reply:
Ah, I see - my mistake!